vllm/model_executor/layers/quantization/fp8.py

```diff
@@ -0,0 +1,148 @@
+import enum
+from enum import Enum
+from typing import Any, Dict, List, Optional
+
+import torch
+from torch.nn.parameter import Parameter
+
+from vllm.model_executor.layers.linear import (LinearMethodBase,
+                                               set_weight_attrs)
+from vllm.model_executor.layers.quantization.base_config import (
+    QuantizationConfig)
+
+
+class FP8Config(QuantizationConfig):
+    """Config class for FP8."""
+
+    @classmethod
+    def get_name(cls) -> str:
+        return "fp8"
+
+    @classmethod
+    def get_supported_act_dtypes(cls) -> List[torch.dtype]:
+        return [torch.bfloat16, torch.half]
+
+    @classmethod
+    def get_min_capability(cls) -> int:
+        # PyTorch 2.3.0+ is required to run FP8 on SM 89 (e.g. Ada) GPUs.
+        # Specifically, this PR has to be included:
+        # https://github.com/pytorch/pytorch/pull/118881/files
+        return 89
+
+    @classmethod
+    def get_config_filenames(cls) -> List[str]:
+        return []
+
+    @classmethod
+    def from_config(cls, config: Dict[str, Any]) -> "FP8Config":
+        return cls()
+
+    def get_linear_method(self) -> "Fp8LinearMethod":
+        return Fp8LinearMethod(self)
+
+    def get_scaled_act_names(self) -> List[str]:
+        return []
+
+
+class Fp8LinearState(Enum):
+
+    UNINITIALIZED = enum.auto()
+    READY = enum.auto()
+
+
+class Fp8LinearMethod(LinearMethodBase):
+    """Linear method for FP8.
+    We now support two types of model checkpoints:
+    1. Common FP16/BF16 model checkpoints. In this case, the weight
+       scaling factor will be initialized during the first forward pass.
+    2. FP8 model checkpoints. In this case, the weight scaling factor
+       will be loaded from the checkpoint with parameter name "w_scale",
+       and the weight should already in FP8 and has been transposed.
+
+    Note that we currently only support per-tensor quantization.
+       
+    Args:
+        quant_config: The quantization config.
+    """
+
+    def __init__(self, quant_config: FP8Config):
+        self.quant_config = quant_config
+
+    def create_weights(
+        self,
+        layer: torch.nn.Module,
+        input_size_per_partition: int,
+        output_size_per_partition: int,
+        input_size: int,
+        output_size: int,
+        params_dtype: torch.dtype,
+        **extra_weight_attrs,
+    ):
+        weight = Parameter(torch.empty(output_size_per_partition,
+                                       input_size_per_partition,
+                                       dtype=params_dtype),
+                           requires_grad=False)
+        set_weight_attrs(weight, {"input_dim": 1, "output_dim": 0})
+        layer.register_parameter("weight", weight)
+        set_weight_attrs(weight, extra_weight_attrs)
+
+        # Will be loaded from FP8 checkpoints, or initialized for
+        # FP16/BF16 checkpoints during the first forward pass.
+        w_scale = Parameter(
```

Comment:
How does the logic for loading this parameter from an FP8 checkpoint work? I thought normally it needs to be done explicitly in the `weight_loader` -- curious how it works since it would also be useful for my fp8 MoE PR :)

Reply:
When loading weights from a checkpoint, we iterate its tensors and assign them to the parameter with the same name. See function load_weights in llama.py for example.

Reply:
Ah got it, the weight loading mechanism only needs to specify the mappings if parameters are renamed / reshuffled, thanks for clarifying, I wasn't aware of this :)

Reply:
We have no way to set it to READY because the logic of creat_weights and loading weights are separated. If FP8 weights are loaded, we simply set to READY in the first forward pass without doing anything.